src/librustc_save_analysis/json_dumper.rs

```diff
@@ -31,8 +29,16 @@ pub struct WriteOutput<'b, W: Write> {
 
 impl<'b, W: Write> DumpOutput for WriteOutput<'b, W> {
     fn dump(&mut self, result: &Analysis) {
-        if write!(self.output, "{}", as_json(&result)).is_err() {
-            error!("Error writing output");
+        let json = match serde_json::to_string(result) {
+            Ok(json) => json,
+            Err(e) => {
+                error!("Can't serialize save-analysis: {}", e);
+                return;
+            }
+        };
+
+        if let Err(e) = write!(self.output, "{}", json) {
```

Comment:
Could use https://docs.serde.rs/serde_json/fn.to_writer.html to avoid intermediate string

Reply:
Unfortunately it only works when passing `impl Write` by value and we have a mutably borrowed one so can't do it without restructuring this a fair bit

Reply:
Write is implemented for $mut impl Write:

https://doc.rust-lang.org/std/io/trait.Write.html#method.by_ref

https://doc.rust-lang.org/src/std/io/impls.rs.html#43-64

Reply:
Ah, I miseed `by_ref`, thanks! Fixed in https://github.com/rust-lang/rust/pull/60053/commits/8498ba204b4c83c21f29d4918057a96cd601baef